Follow up the while signature change.

trentlo opened the pull request:
- Assert that the input and output resources are in the same order to make the
   expected signature of the xla computation more straightforward.

trentlo committed:
Follow up the while signature change (#28092).

 - Assert that the input and output resources are in the same order to make the
   expected signature of the xla computation more straightforward.

trentlo commented:
@sanjoy, this is the (minor) cleanup I promised you a while ago. Please help to take a look.

@jlebar @hawkinsp

sanjoy reviewed (approved):
lgtm

sanjoy reviewed (changes requested):
Hi Trent,

Internally this fails tensorflow/compiler/tests:eager_test_cpu (not sure how it passed presubmit in github, doing an OSS build now).

The problem is that TF eager will sometimes compile regular function calls with XLA, and in such cases the resource outputs may not all be grouped together in the end.  However `XlaComputationLaunchContext::PopulateOutputs` will do the right thing for them.

Do you mind moving this `TF_RET_CHECK` to happen only for while loops since that is the only case where it is guaranteed that this invariant will hold?

trentlo commented:
Understood. Let me see if I can place it in a better place.

trentlo committed:
Move the function signature check to while_op from BuildComputation.

Check that the input resources of the while loop body function are grouped
in the end, as required by the convention of the XLA bridge. We cannot place
this check in BuildComputation as some functions generated by eager mode can
violate it.

trentlo committed:
restored test ReturnResourceHandle.

It tests the reorder behavior of the resource outputs by BuildComputation.

trentlo commented:
Done. Please help to take a look.

Although I think it should not affect the change, I don't fully understand the purpose of the token inputs/outputs. How are they generated, stack access ops?

sanjoy commented on tensorflow/compiler/tf2xla/kernels/while_op.cc:
```diff
@@ -37,6 +37,30 @@ namespace tensorflow {
+    if (has_seen_resource) {
+      if (arg_type != DT_RESOURCE) {
+          return errors::InvalidArgument(
+              "Expect input resources are grouped in the end of while body ",
```
Let's also use `i` in the error message.  If you think that's not useful then please change the loop to a range for loop.

sanjoy commented on tensorflow/compiler/tf2xla/kernels/while_op.cc:
```diff
@@ -37,6 +37,30 @@ namespace tensorflow {
 namespace {
 
+// Verify that input resources are grouped in the end.
+Status VerifyBodyFunctionSignature(XlaOpKernelContext* ctx,
```
Why not `VerifyResourcesArgsGroupedAtEnd`

sanjoy reviewed (changes requested):
Minor nits, otherwise looks great!

sanjoy commented:
> Done. Please help to take a look.
> 
> Although I think it should not affect the change, I don't fully understand the purpose of the token inputs/outputs. How are they generated, stack access ops?

There are some internal users for this, but I'm not sure if there are any external uses.

Roughly: `TOKEN` values let us sequence side effecting operations like `outfeed` and `infeed` in XLA, and we might need to thread these values through function calls, functional while loops etc.

trentlo committed:
Polish function names and error messages.

trentlo commented:
Revised accordingly.

Thanks for the explanations on `token`.

sanjoy reviewed (approved):
lgtm

rthadur commented:
This is merged by the commit: `5205ef9` , so closing this PR
